src/Http/Routing/src/EndpointDataSource.cs

```diff
@@ -22,4 +24,149 @@ public abstract class EndpointDataSource
     /// Returns a read-only collection of <see cref="Endpoint"/> instances.
     /// </summary>
     public abstract IReadOnlyList<Endpoint> Endpoints { get; }
+
+    /// <summary>
+    /// Get the <see cref="RouteEndpoint"/> instances for this <see cref="EndpointDataSource"/> given the specified <see cref="RouteGroupContext.Prefix"/>.
+    /// </summary>
+    /// <param name="context">Details about how the returned <see cref="RouteEndpoint"/> instances should be grouped and a reference to application services.</param>
+    /// <returns>
+    /// Returns a read-only collection of <see cref="RouteEndpoint"/> instances given the specified group <see cref="RouteGroupContext.Prefix"/> and <see cref="RouteGroupContext.Conventions"/>.
+    /// </returns>
+    public virtual IReadOnlyList<Endpoint> GetEndpointGroup(RouteGroupContext context)
```

Comment:
Can this also be `RouteEndpoint` like `RouteEndpointDataSource` is doing?

Reply:
It could, and it originally was, but that's just unnecessarily restrictive.

My reason for originally having it more restrictive is because, the default implementation will throw for any non `RouteEndpoint` anyway, and if you're going to prefix with a `RoutePattern`, don't you want the `Endpoint` type with a `RoutePattern`?

The reason I ultimately decided against it is because you can override the default implementation, and maybe you can meaningfully prefix a custom endpoint or at least apply conventions to it.

The bigger thing is no caller to `GetEndpointGroup` (I still prefer `GetGroupedEndpoints`, but I digress) actually needs the return type to only be `RouteEndpoint`s. That's the biggest reason I decided against it.

I kept `RouteEndpointDataSource` returning `IReadOnlyList<RoutePattern>` because its internal and was a chance to show off an obscure-ish covariant return feature introduced by C# 9, so it doesn't limit any public consumers. It has no practical benefit though.

Reply:
Why does the doc comment say it returns a RouteEndPoint, is that a typo?

Reply:
I also think we should rename this to what it was originally `GetGroupedEndpoints`